bson/unmarshal_test.go

```diff
@@ -166,3 +167,114 @@ func TestCachingDecodersNotSharedAcrossRegistries(t *testing.T) {
 		assert.Equal(t, int32(-1), *second.X, "expected X value to be -1, got %v", *second.X)
 	})
 }
+
+type expectedResponse struct {
+	DefinedField string
+}
+
+func unmarshalExtJSONHelper(extJSON string) (*expectedResponse, error) {
+	responseDoc := expectedResponse{}
+	err := UnmarshalExtJSON([]byte(extJSON), false, &responseDoc)
+	return &responseDoc, err
+}
+
+func TestUnmarshalExtJSONWithUndefinedField(t *testing.T) {
+	// When unmarshalling, fields that are undefined in the destination struct are skipped.
+	// This process must not skip other, defined fields and must not raise errors.
+	testCases := []struct {
+		name     string
+		testJSON string
+	}{
+		{
+			"no array",
+			`{
+				"UndefinedField": {"key": 1},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"outer array",
+			`{
+				"UndefinedField": [{"key": 1}],
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"embedded array",
+			`{
+				"UndefinedField": {"keys": [2]},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"outer array and embedded array",
+			`{
+				"UndefinedField": [{"keys": [2]}],
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"embedded document",
+			`{
+				"UndefinedField": {"key": {"one": "two"}},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"doubly embedded document",
+			`{
+				"UndefinedField": {"key": {"one": {"two": "three"}}},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"embedded document and embedded array",
+			`{
+				"UndefinedField": {"key": {"one": {"two": [3]}}},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"embedded document and embedded array in outer array",
+			`{
+				"UndefinedField": [{"key": {"one": [3]}}],
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"code with scope",
+			`{
+				"UndefinedField": {"logic": {"$code": "foo", "$scope": {"bar": 1}}},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"embedded array of code with scope",
+			`{
+				"UndefinedField": {"logic": [{"$code": "foo", "$scope": {"bar": 1}}]},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"type definition embedded document",
+			`{
+				"UndefinedField": {"myDouble": {"$numberDouble": "1.24"}},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"empty embedded document",
+			`{
+				"UndefinedField": {"empty": {}},
+				"DefinedField": "value"
+			}`,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			responseDoc, err := unmarshalExtJSONHelper(tc.testJSON)
+			require.NoError(t, err)
```

Comment:
Bit of legacy context: many of our older tests use `testify/require` to do assertions, but we're switching to using our own assertion library (`go.mongodb.org/mongo-driver/internal/testutil/assert`). The syntax is very similar but the assertions require a message to go along with them:

```
assert.Nil(t, err, "UnmarshalExtJSON error: %v", err)
assert.Equal(t, "value", responseDoc.DefinedField, "expected DefinedField to be 'value', got %q", responseDoc.DefinedField)
```

This makes them more verbose, but they're clearer because, unlike `testfiy`, they don't rely on the "expected" and "actual" values to be passed in a specific order.

Reply:
Ah gotcha, thanks for the explanation. Switched to use `testutil/assert`.